feat(custom-installers): create a new middle layer for artifacts

Gankra opened the pull request:
* [x] write middle layer
  * [x] cargo-dist
  * [x] inference
  * [x] package-manager  
  * [x] harmonize data sources to feel uniform 
* [x] port downloads table
  * [x] fix `<th>` 1rem padding thing in oranda-css
* [x] port installer widget
   * [x] new JS/html that interprets ReleaseArtifacts (with @shadows-withal)
   * [x] style said html nicely
   * [x] better message for "we actually did detect your OS but no matching platform"
   * [x] add back link to full downloads page
   * [x] add simpler modes for simple cases
 * [x] add back verbose install page listings (use new config)
 * [x] update snapshot tests

Gankra committed:
feat(custom-installers): create a new middle layer for artifacts

Gankra committed:
fixups

manyinsects commented on src/data/artifacts.rs:
```diff
@@ -0,0 +1,235 @@
+    }
+}
+
+/// Take an installer preference an uniformly apply it to every given TargetTriple
```
```suggestion
/// Take an installer preference and uniformly apply it to every given TargetTriple
```

manyinsects reviewed (commented):
looks good as a backbone for this stuff!

Gankra committed:
Update src/data/artifacts.rs

Co-authored-by: liv <shadows@with.al>

Gankra committed:
feat(custom-installers): more progress

This replaces the table rendering code with a new implementation that uses the
new ReleaseArtifacts system. It also does part of the rewiring to make the
installer widget use it too, but currently it just deletes the installer widget
completely, as most of that logic will be moving to JS

Gankra committed:
GC more things out of the cargo_dist module

And rename/fleshout get_os to triple_to_display_name

Commit:
wip: more powerful installer widget

Commit:
fix: consistent comments

Commit:
feat: re-enable autoselection

Commit:
feat: handle nojs cleanly

Commit:
feat: handle autodetection failure cleanly

Gankra committed:
fixups from rebase

Gankra committed:
quick fixes for padding/colors

Gankra committed:
clean up styles a bit

Gankra committed:
add a slightly better message for failed detect

Gankra committed:
make installer widget more simple/robust in simpler cases

Gankra committed:
add back link to page in fallback

Gankra committed:
restore functionality to the artifacts page

Gankra committed:
add back rich view source support

Gankra committed:
fix some tests

Gankra committed:
fix handling of package-managers-only

Gankra committed:
integrate new additional/preferred system

ashleygwilliams commented on src/data/artifact_inference.rs:
```diff
@@ -0,0 +1,352 @@
+
+use super::artifacts::*;
+
+// Architectures
```
are these commented out on purpose?

Gankra replied:
Yeah they might be needed in the future, and wanted to leave them for reference for the js.

ashleygwilliams commented on src/data/artifact_inference.rs:
```diff
@@ -0,0 +1,352 @@
+use camino::Utf8PathBuf;
+
+use super::artifacts::*;
```
i think artifacts should be a dir and this file should be inside it

ashleygwilliams commented on src/data/artifact_inference.rs:
```diff
@@ -0,0 +1,352 @@
+// const OS_MAC: &str = "apple-darwin";
+
+// Various target triples
+const TARGET_X86_WINDOWS: &Targ = "i686-pc-windows-msvc";
```
i think it would be useful to put the constants in a different file potentially

ashleygwilliams commented on src/data/artifact_inference.rs:
```diff
@@ -0,0 +1,352 @@
+            let preference;
+
+            // Try to detect what kind of file this is
+            if file.name.contains("install")
```
could we break these conditions into individual functions?

ashleygwilliams commented on src/data/cargo_dist.rs:
```diff
@@ -1,65 +1,112 @@
+                    self.file_mut(file).checksum_file = checksum_file;
+                }
+
+                match artifact.kind {
```
could we break these out into different functions?

ashleygwilliams commented on src/data/github/mod.rs:
```diff
@@ -74,3 +76,19 @@ impl GithubRelease {
     }
 }
+
+impl ReleaseArtifacts {
```
it is somewhat surprising that this impl is here- i also get why it's here but i wonder if there's not a better way to structure things to make it easier to find

ashleygwilliams commented on src/data/mod.rs:
```diff
@@ -14,43 +16,78 @@ pub struct Context {
     pub releases: Vec<Release>,
     pub has_prereleases: bool,
-    pub latest_dist_release: Option<DistRelease>,
+    pub latest_release: Option<usize>,
```
i am surprised this is a type usize...

Gankra replied:
it's an index into an array

ashleygwilliams replied:
oooooooooooohhhhh- thoughts on call it latest_release_index?

ashleygwilliams commented on src/data/mod.rs:
```diff
@@ -14,43 +16,78 @@ pub struct Context {
-        cargo_dist: bool,
-    ) -> Result<(Vec<Release>, bool, Option<DistRelease>)> {
+        artifacts_config: &ArtifactsConfig,
+    ) -> Result<(Vec<Release>, bool, bool, Option<usize>)> {
```
this return type is starting to look like it wants to be a struct... 2 bools seems like it'll get confusing

Gankra replied:
yeah this function is a disaster

ashleygwilliams commented on src/site/artifacts/installers.rs:
```diff
@@ -1,218 +1,285 @@
+        </div>
+    );
+
+    // If there's only one platform we don't need scripts
```
could we break these out into smaller functions?

ashleygwilliams reviewed (approved):
mostly requests for smaller functions and to re-org some of the file structure in artifacts. we can merge this and i can do that re-org if you'd like

Gankra committed:
address review feedback

ashleygwilliams reviewed (commented):
🚀
